File: app/src/main/java/com/extendvr/BluetoothController.java
package com.extendvr;

import android.bluetooth.BluetoothDevice;
import android.util.Log;

import com.github.douglasjunior.bluetoothclassiclibrary.BluetoothService;
import com.github.douglasjunior.bluetoothclassiclibrary.BluetoothStatus;

public class BluetoothController {
    private BluetoothService service;
    // BluetoothController contains hand and rotation information
    public float w;
    public float x;
    public float y;
    public float z;
    int thumb;
    int indexFinger;
    int middleFinger;
    int ringFinger;
    int pinkie;

    public void connect(BluetoothDevice device){
        // setup tracking callback
        service = BluetoothService.getDefaultInstance();
        service.setOnEventCallback(new BluetoothService.OnBluetoothEventCallback() {
            @Override
            public void onDataRead(byte[] buffer, int length) {
                if (buffer != null && buffer.length > 12) {
                    w = ((buffer[0] << 8) | (buffer[1] & 0xff) )/1000F;
                    x = ((buffer[2] << 8) | (buffer[3] & 0xff) )/1000F;
                    y = ((buffer[4] << 8) | (buffer[5] & 0xff) )/1000F;
                    z = ((buffer[6] << 8) | (buffer[7] & 0xff) )/1000F;
                    thumb = buffer[8] & 0xFF; // actual readings will be between 0 and 99 in final version, for consistency between controllers of
                    // different accuracy, rather than between 0 and 6
                    indexFinger = (buffer[9] & 0xFF) * 99/6;
                    middleFinger = (buffer[10] & 0xFF) * 99/6;
                    ringFinger = (buffer[11] & 0xFF) * 99/6;
                    pinkie = (buffer[12] & 0xFF) * 99/6;
                }
                Log.i("BLUETOOTH_MSG",buffer.toString());
            }

            @Override
            public void onStatusChange(BluetoothStatus status) { }
            @Override
            public void onDeviceName(String deviceName) { }
            @Override
            public void onToast(String message) { }
            @Override
            public void onDataWrite(byte[] buffer) { }
        });


        service.connect(device);
    }
}

File: app/src/main/java/com/extendvr/BluetoothController.kt
package com.extendvr

import android.bluetooth.BluetoothDevice
import android.util.Log
import com.github.douglasjunior.bluetoothclassiclibrary.BluetoothService
import com.github.douglasjunior.bluetoothclassiclibrary.BluetoothService.OnBluetoothEventCallback
import com.github.douglasjunior.bluetoothclassiclibrary.BluetoothStatus

class BluetoothController {
    private var service: BluetoothService? = null

    // BluetoothController contains hand and rotation information, and
    // handles updating based on latest values from the controller
    var w = 0f
    var x = 0f
    var y = 0f
    var z = 0f
    var thumb = 0
    var indexFinger = 0
    var middleFinger = 0
    var ringFinger = 0
    var pinkie = 0
    fun connect(device: BluetoothDevice?) {
        // listen for updates from the controller
        service = BluetoothService.getDefaultInstance()
        service?.setOnEventCallback(object : OnBluetoothEventCallback {
            override fun onDataRead(buffer: ByteArray, length: Int) {
                if (buffer != null && buffer.size > 12) {
                    var data = buffer.toUByteArray()
                    // Get quaternion coordinates from IMU
                    // controller passes them as 2 bit integer 1000 times their actual value
                    // so we convert it to a float with 3dp of precision here
                    w = (data[0] * 256u + data[1]).toFloat()/ 1000f
                    x = (data[2] * 256u + data[3]).toFloat()/ 1000f
                    y = (data[4] * 256u + data[5]).toFloat()/ 1000f
                    z = (data[6] * 256u + data[7]).toFloat()/ 1000f

                    // Get finger readings
                    // Each finger reading is a unsigned byte between 0 and 255
                    // 0 == finger straight up, 255 = finger bent
                    thumb = data[8].toInt()
                    indexFinger = data[9].toInt()
                    middleFinger = data[9].toInt()
                    ringFinger = data[9].toInt()
                    pinkie = data[9].toInt()
                }
                Log.i("BLUETOOTH_MSG", buffer.toString())
            }

            override fun onStatusChange(status: BluetoothStatus) {}
            override fun onDeviceName(deviceName: String) {}
            override fun onToast(message: String) {}
            override fun onDataWrite(buffer: ByteArray) {}
        })
        service?.connect(device)
    }
}

